### DISCUSSION TITLE
queueEvent and pullEvent

### DISCUSSION THREAD

**Discussion by ghost** [cc-tweaked/CC-Tweaked#2023] [Q&A] [answered]
>pullEvent([filter])
>    Pause execution of the current thread and waits for any events matching filter.

>queueEvent(name, ...)
>    Adds an event to the event queue. This event can later be pulled with os.pullEvent.
    
This does not make sense, additionally there does not seem to be a queue.
```lua
lua> os.queueEvent("foo", "bar")
lua> os.pullEvent("foo") -- continues to wait for a event
.
```

**Comment by ghost** [comment]
running the same as a script produces a different result?
```lua
os.queueEvent("foo", "bar")
print(os.pullEvent("foo")) -- > foo bar
```
is there a Time To Live?

**Reply by SquidDev** [reply, marked as answer]
This is expected behaviour, and is because the typing at the Lua prompt causes events to be consumed from the queue. What's actually happening here is:

 - `os.queueEvent("foo", "bar")` queues the `foo` event.
 - The `lua` program calls `read`, which consumes your `foo` event and discards it.
 - You type `os.pullEvent("foo")`. This a dozen `key` and `char` events, which `read` also consumes.
 - `os.pullEvent("foo")` is run. The queue is now empty, so this waits forever.

**Reply by ghost** [reply]
os.sleep() also makes a event?
```lua
os.queueEvent("foo","bar")
os.sleep(0.05)
os.pullEvent("foo") -- waits forever
```
How many events does this "queue" hold? (seems to be one but then `{os.pullEvent()}` returning multiple events wouldn't make sense)
What does call `read` and ruin the queue?

**Reply by SquidDev** [reply]
Yep! You should assume anything which has any delay will wait for an event.

**Reply by ghost** [reply]
So just checking if i understand:
Making a new event destroys previous.
How does `{os.pullEvent()}` return multiple events then?

**Reply by ghost** [reply]
Oh i am stupid, sleep() also reads all previous events and that is what destroys them. Thanks for all the help and swift replies!
